### ISSUE TITLE
a way to indicate a variable is default-initialized?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#15769]
Occasionally it is necessary to avoid split-initialization for one reason or another. For example, a test might need to check the behavior of assignment functions (rather than `init=` functions).

To date, this has involved writing code like
``` chapel
var x: SomeType;
x; // avoid split initialization
```

However this pattern does not work for sync variables (because `x;` would wait for the variable to become full - see #15756). This leads us to thinking of having a way to explicitly request default initialization.

There are also cases where a user might need to explicitly avoid split-initialization to avoid a compilation error. For example, the below program results in compilation error today but will function if `x` is default-initialized.

``` chapel
var x: sync int;
function();
x = 29;

proc function() {
  writeln(x.readXX());
}
```

Lastly, such a functionality would also be useful in the context of initializer arguments for typeless fields (see "Instantiated Type with Two Typeless Fields" in #12318). (Note that supporting typeless fields in this way will require other elements but a strategy here could be suggestive of syntax).

I know of two specific proposals for this feature:

### = init
`var s: someType = init`

Here the compiler would interpret `= init` as explicitly requesting default initialization.

Pros:
 * feels symmetric to `= noinit`
 * `init` is already a special identifier

### = default
`var s: someType = default`

Here the compiler would interpret `= default` as explicitly requesting default initialization.

Pros:
 * naming is relatively obvious

**Comment by mppf** [comment]
I was writing some tests of `=` and it reminded me of this issue.

**Comment by bradcray** [comment]
I still like `= init;` and if a PR showed up implementing it, I'd be a supporter.  I'm still less crazy about `default` primarily because of the introduction of a new special identifier.

**Comment by e-kayrakli** [comment]
` = init` is preferable to me, too. I mostly like the symmetry between `init` and `noinit`.

**Comment by lydia-duncan** [comment]
I'm fine with either, with a slight preference towards `= init`

**Comment by mppf** [comment]
> I still like = init; and if a PR showed up implementing it, I'd be a supporter. I'm still less crazy about default primarily because of the introduction of a new special identifier.

I just realized that, while `noinit` is a keword, `init` is not.

```
var init = 1;
writeln(init);
```

compiles and runs. If we wanted to implement `= init` then presumably `init` would become a keyword. Do people view that as a good or bad outcome?

**Comment by lydia-duncan** [comment]
Does making it a keyword complicate parsing of init functions?

**Comment by mppf** [comment]
> Does making it a keyword complicate parsing of init functions?

Yeah but I don't think such things are insurmountable or anything. At the moment, `init=` is already a keyword (because otherwise it would be two tokens), so making `init` a keyword as well would probably just amount to adjusting the parser for `init` similarly to `init=` in a few places.

**Comment by bradcray** [comment]
I'd consider that a good outcome in that it'd be more similar to `noinit` and `init=`.

**Comment by vasslitvinov** [comment]
+1 to `init` as a keyword. At the very least it should behave like `this` in terms of keyword-ness.

**Comment by bradcray** [comment]
Looking at this issue for the first time in awhile:

* we've decided to move ahead with reserving `init` as part of our strategy of making all special methods either based on reserved words or interfaces (where initializer-related special methods are becoming reserved—init, postinit, deinit) — see https://github.com/chapel-lang/chapel/issues/13897#issuecomment-1603419348
* though I was a big supporter of `= init;` for awhile, I'm a little more gun-shy about it this week for two reasons:

1)  I think that, generally, we want a way to get from a type to its default value, whether that's `type.default`, `defaultOf(type)`, `type.defaultOf()` or what-have-you.  There are cases where we want to be able to refer to this default value that are not necessarily in an initialization setting, like passing the value as an argument to a generic formal.  If we had such a capability, we could write `var x: t = t.default;` or simply `var x = t.default;` rather than needing a new feature in `var x: t = init;` that only works in declaration settings.

2) Recent discussions of `this.complete();` replacements have been considering variations like `this.init;`, `init;`, `this = init;` and if we were to go with an option like that, I'd start to worry that we are leaning on `init` for too many different things (though `this = init;` could potentially be considered to be loosely compatible with the feature request here.

Even if item 2 wasn't here, item 1 feels like a strong argument to me—why introduce a special-case in the language when a more general mechanism that would be useful in other cases achieves the same effect nearly as succinctly?

**Comment by vasslitvinov** [comment]
I like `var x = (some type).default;`

Also, `var x = default(some type);` may be better because it looks consistently with `var x = max(some type);` 

If starting from scratch, I prefer `min`, `max`, and `default` as parenless type methods rather than standalone functions with the type argument.

**Comment by mppf** [comment]
Note that for `myType.default` to work broadly, we have to make `default` a *special method*, right? For this reason, I think it'd be better to have a non-method, like `defaultOf(myType)`.

**Comment by bradcray** [comment]
> we have to make default a special method, right?

Hmm, that's a good point.  Or we'd have to reserve `default`.  

(Just kidding:) Or get it done fast and grandfather it in (which would arguably be safe since it'd be supported for all types?)  

(or am I? :D )

(I am)
